homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,159 @@
+    'pause': STATE_PAUSED,
+    'stop': STATE_IDLE
+}
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    controller = hass.data[HEOS_DOMAIN][DOMAIN]
+    players = controller.get_players()
+    devices = [HeosMediaPlayer(p) for p in players]
+    async_add_devices(devices, True)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, player):
+        """Initialize."""
+        self._player = player
+
+    def _update_state(self):
+        self.async_schedule_update_ha_state()
+
+    async def async_update(self):
+        """Update the player."""
+        self._player.request_update()
+
+    async def async_added_to_hass(self):
+        """Device added to hass."""
+        self._player.state_change_callback = self._update_state
+
+    @property
+    def unique_id(self):
+        """Get unique id of the player."""
+        return self._player.player_id
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._player.name
+
+    @property
+    def volume_level(self):
+        """Volume level of the device (0..1)."""
+        volume = self._player.volume
+        return float(volume) / 100
+
+    @property
+    def state(self):
+        """Get state."""
+        return PLAY_STATE_TO_STATE.get(self._player.play_state)
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def media_content_type(self):
+        """Content type of current playing media."""
+        return MEDIA_TYPE_MUSIC
+
+    @property
+    def media_artist(self):
+        """Artist of current playing media."""
+        return self._player.media_artist
+
+    @property
+    def media_title(self):
+        """Album name of current playing media."""
+        return self._player.media_title
+
+    @property
+    def media_album_name(self):
+        """Album name of current playing media."""
+        return self._player.media_album
+
+    @property
+    def media_image_url(self):
+        """Return the image url of current playing media."""
+        return self._player.media_image_url
+
+    @property
+    def media_content_id(self):
+        """Return the content ID of current playing media."""
+        return self._player.media_id
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._player.mute == 'on'
+
+    async def async_mute_volume(self, mute):
+        """Mute volume."""
+        self._player.set_mute(mute)
+
+    @property
+    def media_duration(self):
+        """Duration of current playing media in seconds."""
+        return self._player.duration / 1000.0
+
+    @property
+    def media_position_updated_at(self):
+        """Get time when position updated."""
+        return self._player.current_position_updated_at
```

Comment:
@awarecan wondering if you can help.  The library sets this to `datetime.now()` whenever `media_position` changes and then raises a callback which causes HASS to update state.  Based on my understanding from the Discord chat, we shouldn't be updating state whenever the playback position changes, which is effectively what is happening here.

What's the right way to implement this functionality?  Heos sends a push update with the playback position as a discrete event -- the values can't be queried and aren't available during other state changes.  It seems we would have to call update state at least one additional time when the first media position is reported after a change in state or song?

Reply:
save last two update (use FIFO queue), compare the incoming status value with last one, if the only change is media position, ignore the HA state update. Always add the new status in the queue, remove the oldest from queue.

Reply:
Thanks for the pointer, the updates for position and other attributes are not differentiated now, so some additional logic or library changes may be required to do this.

For context there's an open arch issue to improve/simplify this for platforms: https://github.com/home-assistant/architecture/issues/101

Reply:
I have a new implementation that hopefully address this. Would be nice to get feedback on this. =)